Initial conversion

garybeverungen opened the pull request:
Closes #1

Commit:
Initial conversion (#1)

First pass at moving keyword screening code from SEER*DMS to reportability-screener

Commit:
Switch to pipe-delimited file (#1)

Remove OpenCSV dependency
Switch keyword file to pipe-delimited
Update file parsing logic

ctmay4 commented on pom.xml:
```diff
@@ -10,7 +10,7 @@
 
     <properties>
         <!-- build under 21 but release as JDK 8 compatible -->
-        <java.version>21</java.version>
```
Why did you change this. It is correct. The `integration.yaml` is building using Java 21 but setting `--release` to Java 8.

```yaml
      - name: Set up JDK
        uses: actions/setup-java@v4
        with:
          distribution: 'adopt'
          java-version: '21'
          cache: 'maven'
```

garybeverungen replied:
Sorry, I guess I misunderstood the emails about this. I thought the takeaway was that we had to build and release using the same java version. I changed it back.

ctmay4 replied:
The discussion was about language level. I was hoping we could use language features from later Java versions and still output Java 8 bytecode. No dice. We have to leave the language level at 8 in IntelliJ.

ctmay4 commented on src/main/java/com/imsweb/ReportabilityScreener.java:
```diff
@@ -1,16 +1,48 @@
+        return result;
+    }
+
+    private void ignoreNegatedPositiveKeywordMatches(List<Keyword> positiveKeywords, List<Keyword> negativeKeywords) {
```
Please add Javadoc explain method.

ctmay4 commented on src/main/java/com/imsweb/ReportabilityScreener.java:
```diff
@@ -1,16 +1,48 @@
+                .forEach(k -> k.setIgnored(true));
+    }
+
+    private ReportabilityResult getResultBasedOnKeywordMatches(List<Keyword> positiveKeywords) {
```
Please add Javadoc explaining method

ctmay4 commented on src/main/java/com/imsweb/ReportabilityScreenerBuilder.java:
```diff
@@ -0,0 +1,86 @@
+        try (BufferedReader reader = new BufferedReader(new FileReader("src/main/resources/default.keyword.list.txt"))) {
+            reader.lines().map(l -> l.split("\\|")).forEach(l -> add(l[0], getGroupFromString(l[1])));
+        }
+        catch (IOException e) {
```
I actually think I would just convert this to an `IllegalStateException` and throw it. This should never happen, but if it does the library is useless.

ctmay4 commented on src/main/java/com/imsweb/ReportabilityScreenerBuilder.java:
```diff
@@ -0,0 +1,86 @@
+
+public class ReportabilityScreenerBuilder {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ReportabilityScreener.class);
```
Sonarqube complains about this. You are using the wrong class for the logger.

garybeverungen replied:
Exceptions are no longer caught and logged, so I ended up removing the LOG altogether.

ctmay4 replied:
Good call. We might want to remove all the logging libraries from the `pom.xml`. Let dependencies the better. Lets wait a bit longer to see if we will want to add logging somewhere else.

ctmay4 commented on src/main/java/com/imsweb/ReportabilityScreenerBuilder.java:
```diff
@@ -0,0 +1,86 @@
+                group = Group.OTHER;
+                break;
+            default:
+                throw new InvalidParameterException("Unexpected group value: " + groupString);
```
`InvalidParameterException` is from java.security. That doesn't seem right. Use `IllegalArgumentException` instead.

ctmay4 commented on src/test/java/com/imsweb/ReportabilityScreenerTest.java:
```diff
@@ -10,8 +12,14 @@
 class ReportabilityScreenerTest {
 
     @Test
```
The tests need to be expanded. You need to have method-based tests for 

- ignoreNegatedPositiveKeywordMatches
- getResultBasedOnKeywordMatches

Also need to have tests that show initializing screening with external keywords.

ctmay4 reviewed (commented):
Great start!

Commit:
Revert change to java.version (#1)

Commit:
Add Javadoc method descriptions (#1)

Commit:
Error handling and logging fixes (#1)

Commit:
Add more unit tests (#1)

- test alternative keyword definition methods
- add method-level tests for ignoreNegatedPositiveKeywordMatches and getResultBasedOnKeywordMatches

Commit:
Fix whole-word-only behavior (#1)

All screening keyword matches should be whole-word-only

Commit:
Remove public modifiers from test methods (#1)

ctmay4 commented:
This is looking good. Can you just clean up the Sonar issues and I think this can be merged.

garybeverungen commented:
SonarCloud is complaining about the use of `.collect(Collectors.toList()));`, but the `.toList()` is not allowed at this language level. I'm not sure how to resolve the issue.

ctmay4 commented:
Interesting. Sonar has the wrong language level. Let me see if I can fix that.

ctmay4 committed:
Set Sonar language level to 8

ctmay4 commented:
I tried to change a setting. If that doesn't work we can just switch the compiler to 8 and be done with it.

ctmay4 commented:
Nevermind. Looks like that fixed it!

ctmay4 commented on src/main/java/com/imsweb/ReportabilityScreener.java:
```diff
@@ -1,16 +1,62 @@
+     */
+    protected ReportabilityResult getResultBasedOnKeywordMatches(List<Keyword> positiveKeywords) {
+        if (positiveKeywords.stream().anyMatch(k -> !k.isIgnored()))
+            return ReportabilityResult.REPORTABLE;
+        else
+            return ReportabilityResult.NON_REPORTABLE;
```
Nitpicky but can you use static imports?

ctmay4 commented on src/main/java/com/imsweb/ReportabilityScreenerBuilder.java:
```diff
@@ -0,0 +1,79 @@
+                group = Group.OTHER;
+                break;
+            default:
+                throw new IllegalArgumentException("Unexpected group value: " + groupString);
```
Code coverage shows you are not testing this.

garybeverungen replied:
Can this go in ReportabilityScreenerTest or should I make a new ReportabilityScreenerBuilderTest class?

ctmay4 replied:
I would create a new test class. That way any additions to the class could be tested there in the future.

ctmay4 commented on src/main/java/com/imsweb/ReportabilityScreenerBuilder.java:
```diff
@@ -0,0 +1,79 @@
+        Group group;
+        switch (groupString) {
+            case "Positive":
+                group = Group.POSITIVE;
+                break;
+            case "Negative":
+                group = Group.NEGATIVE;
+                break;
+            case "Other":
+                group = Group.OTHER;
```
Static imports please.

ctmay4 commented on src/test/java/com/imsweb/ReportabilityScreenerTest.java:
```diff
@@ -10,8 +17,126 @@
+        List<Keyword> negativeKeywordMatches;
+        ReportabilityScreener screener = new ReportabilityScreenerBuilder().build();
+
+        Keyword posKeyword10to20 = new Keyword("posKeyword", 10, 20, Group.POSITIVE);
+        Keyword posKeyword15to25 = new Keyword("posKeyword", 15, 25, Group.POSITIVE);
+        Keyword posKeyword15to16 = new Keyword("posKeyword", 15, 16, Group.POSITIVE);
+        Keyword negKeyword9to21 = new Keyword("negKeyword", 9, 21, Group.NEGATIVE);
+        Keyword negKeyword10to20 = new Keyword("negKeyword", 10, 20, Group.NEGATIVE);
+        Keyword negKeyword15to16 = new Keyword("negKeyword", 15, 16, Group.NEGATIVE);
```
More static imports.

ctmay4 commented:
Just some extremely minor things and then we can merge. Looks good.

Commit:
Use static imports (#1)

Commit:
Add ReportabilityScreenerBuilderTest (#1)

test getGroupFromString()

ctmay4 commented on src/main/java/com/imsweb/ReportabilityScreenerBuilder.java:
```diff
@@ -0,0 +1,79 @@
+    private final TrieBuilder _otherTrieBuilder;
+
+    public ReportabilityScreenerBuilder() {
+        _positiveTrieBuilder = Trie.builder().onlyWholeWords();
+        _negativeTrieBuilder = Trie.builder().onlyWholeWords();
+        _otherTrieBuilder = Trie.builder().onlyWholeWords();
```
I think these are case-sensitive. I assume we want case-insensitive. Please make sure to add unit tests cases to sensitivity and whole words. I actually didn't think we wanted whole words only.

garybeverungen replied:
Good catch. I don't know why I didn't think of that when I made the whole word only change. I'll change that and add unit tests. 

I'm pretty sure we do want whole-word-only. We have very few keywords in SEER*DMS that have whole_word_only=false, and none of them are reportability_alg_keywords.

Commit:
Remove public modifiers (#1)

garybeverungen commented:
Should I add any keyword validation/cleaning to ReportabilityScreenerBuilder.add()?
For example, what should we do if a user tries to add the same keyword twice?
In SEER*DMS, keywords are a max of 200 characters. Should we enforce that?
Should we trim whitespace on the passed keyword?

Commit:
Ignore case when finding keywords (#1)

Commit:
Improve unit tests (#1)

ctmay4 commented:
> Should I add any keyword validation/cleaning to ReportabilityScreenerBuilder.add()? For example, what should we do if a user tries to add the same keyword twice? In SEER*DMS, keywords are a max of 200 characters. Should we enforce that? Should we trim whitespace on the passed keyword?

I'd say yes to all that!

ctmay4 commented on src/main/java/com/imsweb/ReportabilityScreenerBuilder.java:
```diff
@@ -0,0 +1,83 @@
+        }
+    }
+
+    public void add(List<String> keywords, Group group) {
+        keywords.forEach(k -> add(k, group));
+    }
```
I think another potentially useful thing to add to the builder would be `InputStream` and `Reader` add methods. Maybe we could do that in a separate issue if its too much here. The though process is someone could maintain their list in files and those could be loaded in a single command.

ctmay4 replied:
Don’t bother with these new constructors for this issue, we can add it later if needed.

ctmay4 commented on src/main/java/com/imsweb/ReportabilityScreenerBuilder.java:
```diff
@@ -0,0 +1,83 @@
+
+    }
+
+    protected Group getGroupFromString(String groupString) {
+        Group group;
+        switch (groupString) {
+            case "Positive":
+                group = POSITIVE;
+                break;
+            case "Negative":
+                group = NEGATIVE;
+                break;
+            case "Other":
+                group = OTHER;
+                break;
+            default:
+                throw new IllegalArgumentException("Unexpected group value: " + groupString);
```
Also, if people are going to be creating files at some point, we want to be a bit more lenient. This should match us comparing without case. So `POSITIVE` should also work. We might also want to support shorter versions like `POS` or even `P`. Just make it easier for people to maintain these files if they need to.

ctmay4 commented:
@garybeverungen I'd like to merge this soon since i need to work on the Maven Central release process. Do you want to finish this up first or I can merge this and we can create new issues for the few things left.

garybeverungen commented:
I can add the extra validation and loosen the requirements for the getGroupFromString() method. That shouldn't take long. I'll try to have that done today. 
I think the `InputStream` and `Reader` methods can wait.

ctmay4 commented:
I agree that the stream and reader constructors are definitely a future update. Thanks.

Commit:
Relax constraints on keyword list group name (#1)

Commit:
Improve keyword validation/formatting (#1)

Commit:
Remove public modifier (#1)

garybeverungen commented:
I added keyword validation and relaxed keyword group parsing. Let me know if there's anything else to do.

ctmay4 commented:
Looks good. Lets get this merged and I can work on getting the release to work.
